fdbrpc/TokenSign.cpp

```diff
@@ -164,3 +229,193 @@ TEST_CASE("/fdbrpc/TokenSign") {
+	return tmp.append((3 - val.size() % 3) % 3, '=');
+}
+
+template <class GenFun, class ValidateFun>
+void tokenVerifyBench(CryptAlgo algo, GenFun& gen, ValidateFun& validate) {
+	constexpr int numTokens = 10'000;
+	constexpr double testTime = 10;
+	Arena arena;
+	mkcert::KeyPairRef keyPair;
+	switch (algo) {
+	case CryptAlgo::ES256:
+		keyPair = mkcert::KeyPairRef::make(arena);
+		break;
+	case CryptAlgo::HMACSHA256: {
+		// the following is not secure, but good enough for testing
+		// we create a 64 byte key
+		auto k = new (arena) uint8_t[64];
+		static_assert(64 % sizeof(double) == 0);
+		for (int i = 0; i < 64; i += sizeof(double)) {
+			double d = nondeterministicRandom()->random01();
+			memcpy(k + i, &d, sizeof(double));
+		}
+		StringRef key(k, 64);
+		keyPair.privateKeyDer = key;
+		keyPair.publicKeyDer = key;
+		break;
+	}
+	}
+	std::vector<decltype(gen(algo, keyPair.privateKeyDer))> tokens;
+	tokens.reserve(numTokens);
+	fmt::print("Generating tokens");
+	for (int i = 0; i < numTokens; ++i) {
+		tokens.push_back(gen(algo, keyPair.privateKeyDer));
+		if (i % 1000 == 0) {
+			fmt::print(".");
+		}
+	}
+	fmt::print("\n");
+	auto startTime = timer();
+	auto printTime = startTime;
+	double currentTime = startTime;
+	uint64_t iterations = 0;
+	uint64_t lastPrintedIterations = 0;
+	while (startTime + testTime > currentTime) {
+		auto t = tokens[deterministicRandom()->randomInt(0, tokens.size())];
+		validate(algo, t, keyPair.publicKeyDer);
+		if (currentTime - printTime > 1.0) {
+			fmt::print(
+			    "{}s validate {} tokens\n", unsigned(currentTime - startTime), iterations - lastPrintedIterations);
+			lastPrintedIterations = iterations;
+			printTime = currentTime;
+		}
+		currentTime = timer();
+		++iterations;
+	}
+	double endTime = timer();
+	fmt::print("Validated {} tokens in {} seconds ({} tokens per second)\n",
+	           iterations,
+	           endTime - startTime,
+	           unsigned(iterations / (endTime - startTime)));
+}
+
+template <class GenFun, class ValidateFun>
+void tokenVerifyBench(GenFun& gen, ValidateFun& validate) {
+	fmt::print("ES256:\n");
+	fmt::print("======\n");
+	tokenVerifyBench(CryptAlgo::ES256, gen, validate);
+	fmt::print("\n");
+
+	fmt::print("HMAC-SHA256:\n");
+	fmt::print("============\n");
+	tokenVerifyBench(CryptAlgo::HMACSHA256, gen, validate);
+	fmt::print("\n");
+}
+
+Standalone<StringRef> createJWT(CryptAlgo algo,
+                                VectorRef<StringRef> tenants,
+                                double exp,
+                                StringRef keyName,
+                                StringRef key) {
+	rapidjson::StringBuffer headerSS, payloadSS;
+	rapidjson::Writer<rapidjson::StringBuffer> headerW(headerSS), payloadW(payloadSS);
+	headerW.StartObject();
+
+	headerW.String("typ");
+	headerW.String("JWT");
+
+	headerW.String("alg");
+	headerW.String("ES256");
+
+	headerW.EndObject();
+
+	payloadW.StartObject();
+
+	payloadW.String("exp");
+	payloadW.Double(exp);
```

Comment:
Makes sense to make this an integer. At internet scale, sub-second expiry precision is not required and additional precision can needlessly pad the field value (e.g. 10.34562222322)